modules/apps/27-interchain-accounts/types/codec_test.go

```diff
@@ -41,118 +44,291 @@ func (mockSdkMsg) GetSigners() []sdk.AccAddress {
+
+				deserializedMsgs, err := types.DeserializeCosmosTx(simapp.MakeTestEncodingConfig().Marshaler, bz, encoding)
+				if tc.expPass {
+					suite.Require().NoError(err, tc.name)
+				} else {
+					suite.Require().Error(err, tc.name)
+				}
+
+				if tc.expPass {
+					for i, msg := range msgs {
+						// We're using proto.CompactTextString() for comparison instead of suite.Require().Equal() or proto.Equal()
+						// for two main reasons:
+						//
+						// 1. When deserializing from JSON, the `Any` type has private fields and cached values
+						//    that do not match the original message, causing equality checks to fail.
+						//
+						// 2. proto.Equal() does not have built-in support for comparing sdk's math.Int types.
+						//
+						// Using proto.CompactTextString() mitigates these issues by focusing on serialized string representation,
+						// rather than internal details of the types.
+						suite.Require().Equal(proto.CompactTextString(msg), proto.CompactTextString(deserializedMsgs[i]))
+					}
+				}
+			})
+		}
+
+		// test serializing non sdk.Msg type
+		bz, err := types.SerializeCosmosTx(simapp.MakeTestEncodingConfig().Marshaler, []proto.Message{&banktypes.MsgSendResponse{}}, encoding)
+		suite.Require().NoError(err)
+		suite.Require().NotEmpty(bz)
+
+		// test deserializing unknown bytes
+		_, err = types.DeserializeCosmosTx(simapp.MakeTestEncodingConfig().Marshaler, bz, encoding)
+		suite.Require().Error(err) // unregistered type
+
+		// test deserializing unknown bytes
+		msgs, err := types.DeserializeCosmosTx(simapp.MakeTestEncodingConfig().Marshaler, []byte("invalid"), types.EncodingProtobuf)
+		suite.Require().Error(err)
+		suite.Require().Empty(msgs)
+	}
 }
 
 // unregistered bytes causes amino to panic.
 // test that DeserializeCosmosTx gracefully returns an error on
 // unsupported amino codec.
-func (suite *TypesTestSuite) TestDeserializeAndSerializeCosmosTxWithAmino() {
+func (suite *TypesTestSuite) TestProtoDeserializeAndSerializeCosmosTxWithAmino() {
 	cdc := codec.NewLegacyAmino()
 	marshaler := codec.NewAminoCodec(cdc)
 
-	msgs, err := types.SerializeCosmosTx(marshaler, []proto.Message{&banktypes.MsgSend{}})
+	msgs, err := types.SerializeCosmosTx(marshaler, []proto.Message{&banktypes.MsgSend{}}, types.EncodingProtobuf)
 	suite.Require().Error(err)
 	suite.Require().Empty(msgs)
 
-	bz, err := types.DeserializeCosmosTx(marshaler, []byte{0x10, 0})
+	bz, err := types.DeserializeCosmosTx(marshaler, []byte{0x10, 0}, types.EncodingProtobuf)
 	suite.Require().Error(err)
 	suite.Require().Empty(bz)
 }
+
+func (suite *TypesTestSuite) TestUnsupportedEncodingType() {
+	// Test serialize
+	msgs := []proto.Message{
+		&banktypes.MsgSend{
+			FromAddress: TestOwnerAddress,
+			ToAddress:   TestOwnerAddress,
+			Amount:      sdk.NewCoins(sdk.NewCoin("bananas", sdkmath.NewInt(100))),
+		},
+	}
+	_, err := types.SerializeCosmosTx(simapp.MakeTestEncodingConfig().Marshaler, msgs, "unsupported")
+	suite.Require().Error(err)
```

Comment:
nit: add spacing for readability. we can also perform a check on the returned bytes using `Require().Nil(bz)` or `Require().Empty(bz)`, no need to discard

```suggestion
	}
	
	bz, err := types.SerializeCosmosTx(simapp.MakeTestEncodingConfig().Marshaler, msgs, "unsupported")
	suite.Require().Nil(bz)
	suite.Require().Error(err)
```

Reply:
done